File: src/julia/modules/Parameters/src/Parameters.jl
module Parameters

using Random
using Printf
import Enumerations: SolverMethod

include("structures.jl")
include("load_general_configuration.jl")

export ParameterData, read_input_parameters

function save_instance_full_name!(params::ParameterData)::Nothing
	pathSplitted = splitpath(params.inst_path)

	params.name = pathSplitted[end] # e.g.: lr202
	params.type = pathSplitted[end-1] # t1 or t2
	params.group = pathSplitted[end-2] # e.g.: 60R_60V_04I_06M

	if params.cut_off_machs <= 0
		params.cut_off_machs = parse(Int64, params.group[end-2:end-1]) # take the last two digits before 'M', e.g.: 06 from 60R_60V_04I_06M
	end
	params.group = string(params.group[1:end-3], @sprintf("%02d", params.cut_off_machs), "M")

	params.full_name = string(params.name, '_', params.type, '_', params.group)

	println("Instance: ", params.full_name)
	return nothing
end

function read_input_parameters(ARGS::Vector{String})::ParameterData
	println("Running Parameters.read_input_parameters")

	### Set standard values for the parameters ###

	params = ParameterData()
	### Read the parameters and set correct values whenever provided ###
	for param in eachindex(ARGS)
		if ARGS[param] == "--inst_path"
			params.inst_path = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--gen_config_file_path"
			params.gen_config_file_path = ARGS[param+1]
			params.gen_config_file_name = basename(params.gen_config_file_path)[1:end-5]
			load_general_configuration!(params.gen_config_file_path, params)
			param += 1
		elseif ARGS[param] == "--solver"
			params.solver = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--max_time"
			params.max_time = parse(Int, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--print_sol"
			params.print_sol = parse(Int, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--method_type"
			params.method_type = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--method_code"
			params.method_code = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--greedy_service_order"
			params.greedy_service_order = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--cut_off"
			params.cut_off = parse(Int, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--cut_off_machs"
			params.cut_off_machs = parse(Int, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--elevator"
			params.elevator = 1
		elseif ARGS[param] == "--make_instance_feasible"
			params.make_instance_feasible = true
		elseif ARGS[param] == "--epsilon"
			params.epsilon = parse(Float64, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--output"
			params.output = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--suff_outputs"
			params.suff_outputs = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--suff_csv"
			params.suff_csv = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--epsilon_cap"
			params.epsilon_cap = parse(Float64, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--seed"
			params.seed = parse(Int, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--alpha"
			params.alpha = parse(Float64, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--output_flag_grb"
			params.output_flag_grb = parse(Int64, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--mslpr"
			params.mslpr = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--mslpa"
			params.mslpa = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--csv_file_name"
			params.csv_file_name = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--sol_file_name"
			params.sol_file_name = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--timeline_file_name"
			params.timeline_file_name = ARGS[param+1]
			param += 1
		elseif ARGS[param] == "--output_flag_grb_MSLP"
			params.output_flag_grb_MSLP = parse(Int64, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--threads"
			params.threads = parse(Int, ARGS[param+1])
			param += 1
		elseif ARGS[param] == "--solver_method"
			params.solver_method = parse(SolverMethod, ARGS[param+1])
			param += 1
		elseif startswith(ARGS[param], "--")
			error("Unknown parameter ", ARGS[param])
		end
	end

	if !endswith(params.inst_path, '/')
		params.inst_path *= '/'
	end
	if endswith(params.output, '/')
		params.output = params.output[1:end-1]
	end
	params.rng = MersenneTwister(params.seed)
	save_instance_full_name!(params)
	params.csv_file_name = string(
		params.output, "/",
		params.method_type, "_",
		params.method_code, "/",
		"outputs/csvresults",
		"_", params.method_type,
		"_", params.method_code,
		params.suff_csv,
		".csv",
	)
	params.sol_file_name = string(
		params.output, "/",
		params.method_type, "_",
		params.method_code, "/",
		"solutions/",
		params.group, "/",
		params.name, "_sol_",
		params.gen_config_file_name,
		params.suff_outputs,
		".txt",
	)
	params.timeline_file_name = string(
		params.output, "/",
		params.method_type, "_",
		params.method_code, "/",
		"timelines/",
		params.group, "/",
		params.name, "_timeline_",
		params.gen_config_file_name,
		params.suff_outputs,
		".txt",
	)

	return params

end ### end read_input_parameters

end ### end module

File: src/python/modules/parameters/parameters.py
import os
import random

import numpy as np
from .entities import ParameterData
from .load_general_configuration import load_general_configuration

"""
    def parse_constraints_used_mip(params: ParameterData) -> None:
    
    Parse constraints_used_mip_str into a list of bools indicating which constraints are used in the MIP model.
    The input string can contain individual indices or ranges (e.g., "1,2,5-7").
"""
def parse_constraints_used_mip(params: ParameterData) -> None:
    constraints_used_mip_int = []
    if params.constraints_used_mip_str:
        parts = params.constraints_used_mip_str.split(",")
        for part in parts:
            if "-" in part:
                start, end = map(int, part.split("-"))
                constraints_used_mip_int.extend(range(start, end + 1))
            else:
                constraints_used_mip_int.append(int(part))
    max_index = 100 # probably this will never be insufficient
    params.constraints_used_mip = np.zeros(max_index + 1, dtype=bool)
    for c in constraints_used_mip_int:
        if c <= max_index:
            params.constraints_used_mip[c] = True

    return None

def save_instance_full_name(params: ParameterData) -> None:
    path_splitted = os.path.normpath(params.inst_path).split(os.sep)
    params.name = path_splitted[-1]
    params.type = path_splitted[-2]
    params.group = path_splitted[-3]

    if params.cut_off_machs <= 0:
        params.cut_off_machs = int(params.group[-3:-1])

    params.group = f"{params.group[:-3]}{params.cut_off_machs:02d}M"
    params.full_name = f"{params.name}_{params.type}_{params.group}"

    print(f"Instance: {params.full_name}")


def read_input_parameters(args: list[str]) -> ParameterData:
    params = ParameterData()

    i = 1
    while i < len(args):
        if args[i] == "--inst_path":
            params.inst_path = args[i + 1]
            i += 1
        elif args[i] == "--gen_config_file_path":
            params.gen_config_file_path = args[i + 1]
            params.gen_config_file_name = os.path.basename(params.gen_config_file_path)[:-5]
            load_general_configuration(params.gen_config_file_path, params)
            i += 1
        elif args[i] == "--solver":
            params.solver = args[i + 1]
            i += 1
        elif args[i] == "--max_time":
            params.max_time = int(args[i + 1])
            i += 1
        elif args[i] == "--print_sol":
            params.print_sol = int(args[i + 1])
            i += 1
        elif args[i] == "--method_type":
            params.method_type = args[i + 1]
            i += 1
        elif args[i] == "--method_code":
            params.method_code = args[i + 1]
            i += 1
        elif args[i] == "--cut_off":
            params.cut_off = int(args[i + 1])
            i += 1
        elif args[i] == "--cut_off_machs":
            params.cut_off_machs = int(args[i + 1])
            i += 1
        elif args[i] == "--elevator":
            params.elevator = 1
        elif args[i] == "--max_nodes":
            params.max_nodes = int(args[i + 1])
            i += 1
        elif args[i] == "--epsilon":
            params.epsilon = float(args[i + 1])
            i += 1
        elif args[i] == "--output":
            params.output = args[i + 1]
            i += 1
        elif args[i] == "--seed":
            params.seed = int(args[i + 1])
            i += 1
        elif args[i] == "--output_flag_grb_mip":
            params.output_flag_grb_mip = int(args[i + 1])
            i += 1
        elif args[i] == "--mip_grb_max_time":
            params.mip_grb_max_time = int(args[i + 1])
            i += 1
        elif args[i] == "--mip_hx_max_time":
            params.mip_hx_max_time = int(args[i + 1])
            i += 1
        elif args[i] == "--csv_file_name":
            params.csv_file_name = args[i + 1]
            i += 1
        elif args[i] == "--sol_file_name":
            params.sol_file_name = args[i + 1]
            i += 1
        elif args[i] == "--timeline_file_name":
            params.timeline_file_name = args[i + 1]
            i += 1
        elif args[i] == "--grb_file_name":
            params.grb_file_name = args[i + 1]
            i += 1
        elif args[i] == "--validate_synchronization":
            params.validate_synchronization = True
        elif args[i] == "--constraints_used_mip_str":
            params.constraints_used_mip_str = args[i + 1]
            i += 1
        elif args[i] == "--run_callback_mip_gurobi":
            params.run_callback_mip_gurobi = True
            i += 1
        elif args[i] == "--mip_grb_heuristics":
            params.mip_grb_heuristics = float(args[i + 1])
            i += 1
        elif args[i] == "--threads":
            params.threads = int(args[i + 1])
            i += 1
        # Add other parameters as needed
        elif args[i].startswith("--"):
            raise ValueError(f"Parameter {args[i]} does not exist")
        i += 1

    if not params.inst_path.endswith("/"):
        params.inst_path += "/"

    params.gen_config_file_name = os.path.basename(params.gen_config_file_path)[:-5]
    params.rng = random.Random(params.seed)
    save_instance_full_name(params)
    print(params)
    if not params.output.endswith("/"):
        params.output += "/"
    params.csv_file_name = f"{params.output}{params.method_type}_{params.method_code}/outputs/csvresults_{params.method_type}_{params.method_code}.csv"
    params.sol_file_name = f"{params.output}{params.method_type}_{params.method_code}/solutions/{params.group}/{params.name}_sol_{params.gen_config_file_name}.txt"
    params.timeline_file_name = f"{params.output}{params.method_type}_{params.method_code}/timelines/{params.group}/{params.name}_timeline_{params.gen_config_file_name}.txt"
    params.grb_file_name = f"{params.output}{params.method_type}_{params.method_code}/gurobi/{params.group}/{params.name}_grb_{params.gen_config_file_name}.log"
    parse_constraints_used_mip(params)
    print("Constraints active in MIP")
    for i in range(1,50):
        if params.constraints_used_mip[i]:
            print(f"{i}", end=" ")
            
    print()
    return params